Exception-Handling.md

```diff
@@ -0,0 +1,254 @@
+
+So in the following example:
+
+```ruby
+def quotient_and_remainer dividend, divisor
+  answer = [dividend / divisor]
+  answer << dividend % divisor
+  answer
+end
+
+def mystery(num)
+  answer = []
+
+  (0...9).each do |i|
+    answer << quotient_and_remainer(num, i)
+  end
+  answer
+end
+
+begin
+  puts "The Remainders are #{mystery(100)}"
+end
+```
+
+1.  Ruby Starts by calling the `mystery` method which starts a loop with values from 0 to 9 (inclusive).  
+2.  It then calls the `quotient_and__remainer` method with 100 & 0 as arguments.  
+3.  Inside the `quotient_and_remainder` method it has a `ZeroDivisionError` error.
+4.  Since the error is not rescued in `quotient_and_remainer` Ruby falls back to mystery.
+5.  Since the error is not rescued in `mystery` Ruby falls back to the main program and exits with an error.
+
+You can see the results in the error message:
+
+```bash
+multimethod.rb:2:in `/': divided by 0 (ZeroDivisionError)
+	from multimethod.rb:2:in `quotient_and_remainer'
+	from multimethod.rb:11:in `block in mystery'
+	from multimethod.rb:10:in `each'
+	from multimethod.rb:10:in `mystery'
+	from multimethod.rb:17:in `<main>'
+```
+
+As you can see the error messages display in order as Ruby falls back from where it encountered the error, moving, "through the stack," unwinding the method calls that lead to the error.  
+
+
+
+#### Exception Handling in Rails
+
+In Rails we have a number of methods in ActiveRecord which can generate exceptions a few of which include:
+-  save!
+-  create!
+-  update!
+
+These methods trigger exceptions when validations fail or the database is unavailable.  So you could handle problems with creating a new task in TaskListRails using exceptions as follows.  
+
+```ruby
+  def create
+    @task = Task.new(task_params)
+
+    begin
+      @task.save!
+    rescue ActiveRecord::RecordInvalid
+      flash[:notice] = "Cannot Save the Task."
+    ensure
+      redirect_to tasks_new_path
+    end
+  end
+```
+
+However while this can be done it isn't good practice.  It's considered best practice to [**not use exceptions for expected outcomes**](https://robots.thoughtbot.com/save-bang-your-head-active-record-will-drive-you-mad).  Since we expect users to give us invalid input on a regular basis, the above example does not qualify.  This leaves us with the question:  
+
+### Why Would We Use Rescue?
+
+There are a lot of [differing opinions](https://www.sitepoint.com/ruby-error-handling-beyond-basics/) on when and why you should implement exception handling in your code.  The consensus seems to be that exception handling should be for *exceptional* circumstances.  In other words you use Exception Handling (rescue) for abnormal events, and regular if statements for normal occurrences like validation errors.
+
+You can think of your program errors falling into two categories:
+-  Things you expect to happen regularly, like:
```

Comment:
Since we're essentially replicating this list below, I think we can just take out this portion and leave the second version, which is more detailed.